### PR TITLE
feat: add automatic timestamps for various tables

### PR DESCRIPTION
**Pull request by iPurpl3x**
**Changes proposed in this pull request:**
Database migrations that add timestamps to various tables: 
- `groups`
- `group_user`
- `group_permission`
- `tags`
- `discussion_tag`
- `post_mentions_post`
- `post_mentions_user`

The timestamps are not really used anywhere in Flarum, but can be very handy for analytics purposes.

This PR is kind of a follow-up of https://github.com/flarum/likes/pull/28

**Reviewers should focus on:**
The migrations run without error and don't cause any unwanted side-effects (including performance aspects).

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

### CODE REVIEW

**Review by askvortsov1 — APPROVED** [review]
Haven't tested locally, but makes sense!

**Review by SychO9 — COMMENTED** [review]
Could you please for every timestamp add it to the relevant model's `$dates` array property for proper casting?

also very minor but there seems to be a space in the filename:
`2022_05_20_000005_add_created_at_to_ post_mentions_post_table.php`

**Comment by davwheat** [comment]
The only issue with this is that it would set `created_at` to the time the migration is run, which isn't accurate and could pollute statistics.

Would it not be better to set to null initially, then update the default value to be the current time?

**Comment by iPurpl3x** [comment]
@SychO9 Sure, I added the casts (for the tables that have corresponding models) and fixed the file name (good that you saw that one!)

@davwheat I have thought of that. The thing is that:

_IF_ we choose to update the value of the column automatically in the DB (and not use Laravel, which might slightly impact performance) then the `DEFAULT` is going to be `current_timestamp()` so I would need to run a second migration that sets all the old values to `NULL` after the column(s) have been added. Do you want me to do that ?

**Comment by davwheat** [comment]
In my opinion, running two migrations to do that (can they be merged into two steps of one migration? I have no idea) would be the best option.

I'd be interested to hear the opinions of other maintainers though.

**Comment by iPurpl3x** [comment]
@davwheat If I don't use the `\Flarum\Database\Migration` helper, and write the migration by hand, I might be able to merge it all into one file. 

I also would like to hear opinions of others...

**Comment by SychO9** [comment]
How about creating the columns first, then modifying them to use current (and use current on update where appropriate). That avoids having to modify all rows.

```php
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Database\Schema\Builder;

return [
    'up' => function (Builder $schema) {
        $schema->table('groups', function (Blueprint $table) {
            $table->timestamp('created_at')->nullable();
        });

        // do this manually because dbal doesn't recognize timestamp columns
        $connection = $schema->getConnection();
        $prefix = $connection->getTablePrefix();
        $connection->statement("ALTER TABLE $prefix`groups` MODIFY created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP");
    },

    'down' => function (Builder $schema) {
        $schema->table('groups', function (Blueprint $table) {
            $table->dropColumn('created_at');
        });
    }
];
```

Also it might be good to have a comment on each migration that mentions the columns were added for third-party analytical purposes.

**Comment by iPurpl3x** [comment]
Ah yes, that's pretty clever @SychO9, I will try that I keep you updated.

Agreed for the comments.

**Comment by iPurpl3x** [comment]
I have now updated all the migrations, using the idea of @SychO9 to makes that all existing rows in the DB stay with NULL on the timestamps and don't get a false value of the time when the migration is done.

Hopefully that makes this PR now mergeable 🤞🏻

**Review by davwheat — APPROVED** [review]
Tested locally. Seems to work as intended on non-prefixed DBs.

image
image

File: extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+return [
+    'up' => function (Builder $schema) {
+        $schema->table('post_mentions_post', function (Blueprint $table) {
+            $table->timestamp('created_at')->nullable();
+        });
+
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
Comment (davwheat) [inline extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php]: As per tests failing, should this (and others) instead be....

```suggestion
        $connection->statement("ALTER TABLE `${prefix}post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

Comment (SychO9) [reply extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php]: yup, my bad

```suggestion
        $connection->statement("ALTER TABLE `{$prefix}post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

Comment (davwheat) [reply extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php]: I'll go through and update them -- no need to inconvenience Rafael with that :)

Comment (davwheat) [inline extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php]:
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

File: extensions/mentions/migrations/2022_05_20_000006_add_created_at_to_post_mentions_user_table.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+return [
+    'up' => function (Builder $schema) {
+        $schema->table('post_mentions_user', function (Blueprint $table) {
+            $table->timestamp('created_at')->nullable();
+        });
+
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`post_mentions_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
Comment (davwheat) [inline extensions/mentions/migrations/2022_05_20_000006_add_created_at_to_post_mentions_user_table.php]:
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}post_mentions_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

File: extensions/tags/migrations/2022_05_20_000003_add_timestamps_to_tags_table.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        $schema->table('tags', function (Blueprint $table) {
+            $table->timestamp('created_at')->nullable();
+            $table->timestamp('updated_at')->nullable();
+        });
+
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`tags` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
+        $connection->statement("ALTER TABLE $prefix`tags` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```
Comment (davwheat) [inline extensions/tags/migrations/2022_05_20_000003_add_timestamps_to_tags_table.php]:
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}tags` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
        $connection->statement("ALTER TABLE `{$prefix}tags` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```

File: extensions/tags/migrations/2022_05_20_000004_add_created_at_to_discussion_tag_table.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+return [
+    'up' => function (Builder $schema) {
+        $schema->table('discussion_tag', function (Blueprint $table) {
+            $table->timestamp('created_at')->nullable();
+        });
+
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`discussion_tag` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
Comment (davwheat) [inline extensions/tags/migrations/2022_05_20_000004_add_created_at_to_discussion_tag_table.php]:
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}discussion_tag` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

File: framework/core/migrations/2022_05_20_000000_add_timestamps_to_groups_table.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        $schema->table('groups', function (Blueprint $table) {
+            $table->timestamp('created_at')->nullable();
+            $table->timestamp('updated_at')->nullable();
+        });
+
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`groups` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
+        $connection->statement("ALTER TABLE $prefix`groups` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```
Comment (davwheat) [inline framework/core/migrations/2022_05_20_000000_add_timestamps_to_groups_table.php]:
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}groups` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
        $connection->statement("ALTER TABLE `{$prefix}groups` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```

File: framework/core/migrations/2022_05_20_000001_add_created_at_to_group_user_table.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+return [
+    'up' => function (Builder $schema) {
+        $schema->table('group_user', function (Blueprint $table) {
+            $table->timestamp('created_at')->nullable();
+        });
+
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`group_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
Comment (davwheat) [inline framework/core/migrations/2022_05_20_000001_add_created_at_to_group_user_table.php]:
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}group_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

File: framework/core/migrations/2022_05_20_000002_add_created_at_to_group_permission_table.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+return [
+    'up' => function (Builder $schema) {
+        $schema->table('group_permission', function (Blueprint $table) {
+            $table->timestamp('created_at')->nullable();
+        });
+
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`group_permission` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
Comment (davwheat) [inline framework/core/migrations/2022_05_20_000002_add_created_at_to_group_permission_table.php]:
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}group_permission` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

File: extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php (hunk 1 above)
Comment (davwheat) [reply extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php]: Oh yeah, I don't have permission to push to GB's fork! 🙈 

The suggestions are all there for @iPurpl3x to bulk-commit anyway :)

Comment (iPurpl3x) [reply extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php]: Done @davwheat

**Review by davwheat — APPROVED** [review]

**Review by SychO9 — APPROVED** [review]

**Pull request merged**